primitives/runtime/src/generic/digest.rs

```diff
@@ -82,10 +82,14 @@ impl<Hash> Digest<Hash> {
 #[cfg_attr(feature = "std", derive(parity_util_mem::MallocSizeOf))]
 pub enum DigestItem<Hash> {
 	/// System digest item that contains the root of changes trie at given
-	/// block. It is created for every block iff runtime supports changes
+	/// block. It is created for every block if runtime supports changes
 	/// trie creation.
 	ChangesTrieRoot(Hash),
 
+	/// Digest item that signals to light clients that :code or :heappages
+	/// has changed enabling them request the new code/heappages from full nodes.
+	RuntimeCodeChanged(Hash, u64),
```

Comment:
One of the dangers here is that there is no guarantee that a runtime will actually emit this digest item. Or that the runtime may emit this digest item wrongly. The client code has to account for these possibilities, or explicitly clarify its expectations.

Reply:
Anyway, can't light clients just track `:code` or` :heappages` using the `ChangesTrieRoot`?

Reply:
> Anyway, can't light clients just track :code or :heappages using the ChangesTrieRoot?

but isn't this for runtimes that support emitting the changes trie? but yeah I agree, it'd be more idiomatic.

Reply:
Part of my point is that your code is only for runtimes that support the `RuntimeCodeChanged` protocol exactly as described.

And that running the light client on arbitrary chains will not yield expected results - since they might not issue `RuntimeCodeChanged` digests at the correct points (what happens then? light client breaks?) or even issue them wrongly (same questions arise).

Reply:
ahh, so how do runtimes signal what digest items they support?

Reply:
I don't think we have a signaling mechanism. It's something that would be important at genesis, but that cat's out of the bag on some networks already :) 

Malicious runtime is not something that I would be too worried about, but note that also for those same networks, they aren't issuing the `RuntimeCodeChanged` digest until later on, so the light client wouldn't function until fully synced.

I just expect these assumptions to be documented within the implementation of the light client protocol with the understanding that upgradeable chains can break light clients if they introduce a runtime which doesn't follow the expected rules